pkg/ingester/transfer.go

```diff
@@ -0,0 +1,235 @@
+		// chunkSet.FromIngesterId and assume it is the same every time around
+		// this loop.
+		if fromIngesterID == "" {
+			fromIngesterID = chunkSet.FromIngesterId
+			level.Info(util.Logger).Log("msg", "processing TransferChunks request", "from_ingester", fromIngesterID)
+		}
+
+		userCtx := user.InjectOrgID(stream.Context(), chunkSet.UserId)
+
+		lbls := []client.LabelAdapter{}
+		for _, lbl := range chunkSet.Labels {
+			lbls = append(lbls, client.LabelAdapter{Name: lbl.Name, Value: lbl.Value})
+		}
+
+		instance := i.getOrCreateInstance(chunkSet.UserId)
+		for _, chunk := range chunkSet.Chunks {
+			if err := instance.consumeChunk(userCtx, lbls, chunk); err != nil {
+				return err
+			}
+		}
+
+		seriesReceived++
+		receivedChunks.Add(float64(len(chunkSet.Chunks)))
+	}
+
+	if seriesReceived == 0 {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no series", "from_ingester", fromIngesterID)
+		return fmt.Errorf("no series")
+	} else if fromIngesterID == "" {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no ID from ingester")
+		return fmt.Errorf("no ingester id")
+	}
+
+	if err := i.lifecycler.ClaimTokensFor(stream.Context(), fromIngesterID); err != nil {
+		return err
+	}
+
+	if err := i.lifecycler.ChangeState(stream.Context(), ring.ACTIVE); err != nil {
+		return err
+	}
+
+	// Close the stream last, as this is what tells the "from" ingester that
+	// it's OK to shut down.
+	if err := stream.SendAndClose(&logproto.TransferChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing TransferChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+	level.Info(util.Logger).Log("msg", "Successfully transferred chunks", "from_ingester", fromIngesterID, "series_received", seriesReceived)
+	return nil
+}
+
+// StopIncomingRequests implements ring.Lifecycler.
+func (i *Ingester) StopIncomingRequests() {
+	i.instancesMtx.Lock()
+	defer i.instancesMtx.Unlock()
+	i.readonly = true
+}
+
+// TransferOut implements ring.Lifecycler.
+func (i *Ingester) TransferOut(ctx context.Context) error {
+	backoff := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 5 * time.Second,
+		MaxRetries: i.cfg.MaxTransferRetries,
+	})
+
+	for backoff.Ongoing() {
+		err := i.transferOut(ctx)
+		if err == nil {
+			return nil
+		}
+
+		level.Error(util.Logger).Log("msg", "transfer failed", "err", err)
+		backoff.Wait()
+	}
+
+	return backoff.Err()
+}
+
+func (i *Ingester) transferOut(ctx context.Context) error {
+	targetIngester, err := i.findTransferTarget(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot find ingester to transfer chunks to: %v", err)
+	}
+
+	level.Info(util.Logger).Log("msg", "sending chunks", "to_ingester", targetIngester.Addr)
+	c, err := i.cfg.ingesterClientFactory(i.clientConfig, targetIngester.Addr)
+	if err != nil {
+		return err
+	}
+	if c, ok := c.(io.Closer); ok {
+		defer c.Close()
```

Comment:
there a util function you can use to log and close. `helpers.LogError("closing client", c.Close)`

Reply:
ah, that's cool, done